### ISSUE TITLE
Unclear compilation error "cannot define variable _ inside and".

### ISSUE THREAD

**Issue by TBK145** [elixir-lang/elixir#5256]
### Environment

* Elixir version (elixir -v):

Erlang/OTP 19 [erts-8.1] [source] [64-bit] [smp:4:4] [async-threads:10] [kernel-poll:false]

IEx 1.4.0-dev (662ecf9)

* Operating system:
Ubuntu 16.04

### Current behavior

When compiling this code snippet:
```
        if response.result == :failure and
           response.content.code in 300..399 and
           cmd.state.recurse do
```

I get the following error (which also does not show a line number, so it took me quite some time to find the offending statement):
```
== Compilation error on file lib/dialplan/command/proxy.ex ==
** (CompileError) lib/dialplan/command/proxy.ex: cannot define variable _ inside and
    (stdlib) lists.erl:1338: :lists.foreach/2
    (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
```

On Elixir 1.3.3 it compiles fine, and if I change the code to either
```
        if response.result == :failure and
           response.content.code >= 300 and
           response.content.code <= 400 and
           cmd.state.recurse do
```
or
```
        code = response.content.code
        if response.result == :failure and
           code in 300..399 and
           cmd.state.recurse do
```
It compiles without errors.

### Expected behavior

I expect this code to be compiled, just like on 1.3.3.

**Comment by TBK145** [comment]
I don't know if this bug still exists, but compilation is fixed on 819d81c.

**Comment by josevalim** [comment]
@lexmag given the other commit also fixed the bug, should we abort the changes to and/or and directly add the test?

**Comment by lexmag** [comment]
Unfortunately that commit doesn't fix the test in #5258.

**Comment by lexmag** [comment]
Closed via #5295.
